main/src/ca/uwaterloo/flix/api/lsp/provider/Visitor.scala

```diff
@@ -0,0 +1,850 @@
+    * @param acceptor  An acceptor which defines the the criteria for whether an AST node should be visited.
+    */
+  def visitRoot(root: Root, consumer: Consumer, acceptor: Acceptor): Unit = {
+
+    implicit val c = consumer
+    implicit val a = acceptor
+
+    root.defs.foreach{ case (_, defn) => visitDef(defn) }
+
+    root.effects.foreach{ case (_, eff) => visitEffect(eff) }
+
+    // root.entryPoint.map{ case v => visitEntryPoint(visit, accept)(v) }
+
+    root.enums.foreach{ case (_, e) => visitEnum(e) }
+
+    root.instances.foreach{ case (_, l) => l.foreach(visitInstance) }
+
+    // root.modules.map { case (_, v) => visitModule(v, ???, ???) }
+
+    root.sigs.foreach{ case (_, sig) => visitSig(sig) }
+
+    root.structs.foreach{ case (_, struct) => visitStruct(struct) }
+
+    // root.traitEnv.map{ case (_, v) => visitTraitEnv(???, ???)(v) };
+
+    root.traits.foreach{ case (_, traitt) => visitTrait(traitt) }
+
+    root.typeAliases.foreach{ case (_, alias) => visitTypeAlias(alias) }
+
+    // root.uses
+  }
+
+  def inside(uri: String, pos: Position)(loc: SourceLocation): Boolean = {
+    val (x, y) = pos.toZeroIndexed
+
+    val posLine = x + 1
+    val posCol = y + 1
+
+    // sp1 and sp2, by invariant, has the same source, so we can use either
+    (uri == loc.sp1.source.name) &&
+    (posLine >= loc.beginLine) &&
+    (posLine <= loc.endLine) &&
+    (!(posLine == loc.beginLine && posCol < loc.beginCol)) &&
+    (!(posLine == loc.endLine && posCol >= loc.endCol)) // geq because end column is exclusive
+  }
+
+  def inside(loc1: SourceLocation, loc2: SourceLocation): Boolean = {
+    loc1.source == loc2.source &&
+    (loc1.beginLine >= loc2.beginLine) &&
+    (loc1.endLine <= loc2.endLine) &&
+    !(loc2.beginLine == loc1.beginLine && loc2.beginCol > loc1.beginCol) &&
+    !(loc1.endLine == loc2.endLine && loc1.endCol > loc2.endCol)
+  }
+
+  private def visitEnum(enm: Enum)(implicit a: Acceptor, c: Consumer): Unit = {
+    if (!a.accept(enm.loc)) { return }
+
+    c.consumeEnum(enm)
+
+    visitAnnotations(enm.ann)
+    // TODO visit modifiers?
+    enm.tparams.foreach(visitTypeParam)
+    visitDeriveList(enm.derives)
+    enm.cases.map(_._2).foreach(visitCase)
+  }
+
+  private def visitDeriveList(deriveList: Derivations)(implicit a: Acceptor, c: Consumer): Unit = {
+    if (!a.accept(deriveList.loc)) { return }
+    c.consumeDeriveList(deriveList)
+    deriveList.traits.foreach(visitDerive)
+  }
+
+  private def visitDerive(derive: Derivation)(implicit a: Acceptor, c: Consumer): Unit = {
+    if (!a.accept(derive.loc)) { return }
+    c.consumeDerive(derive)
+  }
+
+  private def visitCase(cse: Case)(implicit a: Acceptor, c: Consumer): Unit = {
+    if (!a.accept(cse.loc)) { return }
+    c.consumeCase(cse)
+    // TODO visit scheme?
+  }
+
+  private def visitInstance(ins: Instance)(implicit a: Acceptor, c: Consumer): Unit = {
+    if (!a.accept(ins.loc)) { return }
+
+    visitAnnotations(ins.ann)
```

Comment:
You know, if you pattern match on Instance then you can see what components are there.

This would make it much easier to immediately see if you have visited everything.

Reply:
Moreover, if the definition of Instance changes then this code would break, but your code would not.

Reply:
So I suggest to use pattern matching everywhere, i.e. def visitX() should start with a pattern match on `X`

Reply:
Good point. I think I'll do that